OTEP: Span Event API deprecation plan (#4430)
open-telemetry/opentelemetry-specification: oteps/4430-span-event-api-deprecation-plan.md

# Span Event API deprecation plan

## Motivation

[OTEP 265: Event Vision](0265-event-vision.md) states that we intend to
deprecate Span Events in favor of (log-based) Events.

After further discussions, we are only planning to deprecate the Span Event
API, while retaining the ability to emit Span Events via the Logs API.

This achieves the primary goal of deprecation, which is to provide a single
consistent guidance to instrumentation authors to use the Logs API when
emitting events, while still supporting use cases that rely on Span Events
being emitted in the same proto envelope as their containing span.

## The plan

### In the proto

Stabilize (log-based) Events.

### In the Specification

1. Stabilize emitting exceptions and events via the Logs API.

   This will allow recording exceptions and events using the Logs API,
   instead of recording them using the Span Event API.

> 2. Mark [Span RecordException](../specification/trace/api.md#record-exception)

Reviewer: Based on the discussion in https://github.com/open-telemetry/opentelemetry-specification/issues/4429#issue-2874711609, the implication here is that for users who haven't configured logging, exceptions (particularly stacktraces) which were previously recorded on spans will instead not be recorded at all, as code moves away from `RecordException`.

Moreover, 'typical' configuration of logs won't attach them to spans, so even when exceptions are recorded they will be moved away from the spans that they ended.

These seem like major decisions that should be explained and justified in this OTEP.

Reviewer: It's also still not clear to me how the Python SDK (or [Ruby](https://github.com/open-telemetry/opentelemetry-ruby/blob/aacd8c8e264b110507c2a733dcc5309ca26aac66/api/lib/opentelemetry/trace/tracer.rb#L34)) which has commonly used and recommended conveniences for recording exceptions on spans (using only the tracing SDK) will function.

https://github.com/open-telemetry/opentelemetry-specification/pull/4430#discussion_r2008804689 by @lmolkova is not really an answer. `setException` was in this PR and was removed. Even if it could be added in a future OTEP, the implication of accepting this OTEP is "this is an OK path forward *even if we never add `setException` or similar in the future". If adding something in the future is the solution then that needs to be explicit here.

Author: discussed with @alexmojaki and he is ok to resolve this discussion

   recommending instead that events are recorded using the Logs API.

   This can be done in parallel with 2.

### Per API and SDK

1. Implement and stabilize emitting exceptions and events via the Logs API.

2. Implement the SDK-based backcompat story:

   - [Sending (log-based) Events as Span Events](#via-the-sdk)

3. Mark
   [Span RecordException](../specification/trace/api.md#record-exception)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that exceptions are recorded using the Logs API.

4. Mark [Span AddEvent](../specification/trace/api.md#add-events)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that events are recorded using the Logs API.

   This can be done in parallel with 3.

### Per Instrumentation

For [stable](../specification/versioning-and-stability.md#stable)
instrumentations that are using
[Span RecordException](../specification/trace/api.md#record-exception)
and [Span AddEvent](../specification/trace/api.md#add-events):

- In the instrumentation's current major version
  - It SHOULD continue to use these
- In the instrumentation's next major version

>   - It SHOULD stop using these,

Reviewer 2: This should have been the case when modelling before, right?

What's an example of these?

Author: I'm guessing you meant to comment on the line below?

> instead recording additional details about spans (that don't need a timestamp) as spans attributes

if so, hopefully this helps: https://github.com/open-telemetry/opentelemetry-specification/pull/4430/commits/b7d37fd62b260a9fa281d319360c109c8d4d7c62

    instead recording additional details about spans (that don't need a timestamp)
    as spans attributes,
    and recording exceptions and events using the Logs API.
  - Users will be able to retain the prior telemetry output by opting in to
    - [Sending (log-based) Events as Span Events](#sending-log-based-events-as-span-events)

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance.

## Sending (log-based) Events as Span Events

### Via the SDK

> There MUST be a way to send (log-based) Events as Span Events

Reviewer: Suppose I configure this. Then some time later I want to start using log-events that aren't emitted within the context of a span. So I need to configure logs to actually be sent via OTLP. Does this mean I also have to start sending new logs that were previously attached to a span?

Author: > Does this mean I also have to start sending new logs that were previously attached to a span?

I'm not sure I follow this question, can you elaborate? thanks

Reviewer 3: A agree with @alexmojaki here - there are a lot of logs reported in scope of a span, and only a few (if any) would be reported as span events. 

By deprecating Span Events, we remove the ability to say that 'this event belongs to that span, it should be sampled and visualized along with it' and separate it from all the other logs.

It can be addressed in a few different ways:

1. a convention: E.g. if Logger instrumentation scope is the same as Tracer instrumentation scope, these logs and spans came from the same source, it can be used as a signal to attach them to span. 
2. explicit API parameter on the `LoggerProvider.getLogger` allowing to request logger that emits this special class of events associated with a span.
3. explicit API parameter when emitting a log advising that it belongs with a span (it may also be useful on backends to highlight or prioritize showing them along with a span)
4. a more complicated convention: span adding a reference to the event via an attribute. E.g. when recording exception, log record is enriched with `exception.id` and the span gets the same `exception.id` attribute.  This covers some cases, but not a general solution.

My preference would be to start with 1. I believe that backends that don't support logs is a temporary problem.

Having a convention allows us to solve the temporary problem without introducing anything new. If it turns our that I'm wrong we can always add new APIs at that point.

Reviewer 3: An easy solution we discussed in the Logs SIG 4/8 is that we'd only convert things that have non-empty EventName to Span events and someone can build additional (custom) filtering.

Given that this OTEP proposes converting just events (not all logs) to span events, it should not significantly change any consumption experience, so I'm good with the current state of the OTEP

Author: I'm resolving this discussion, but if I'm mistaken, please open a new discussion summarizing the remaining open discussion points, thanks!

for use cases that rely on Span Events being emitted in the
same proto envelope as their containing span, and for users
who need to preserve the prior behavior when updating to
new instrumentation.

This mechanism SHOULD be implemented as follows (see
[prototype](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/80adbe1cf8de647afa32c68f921aef2bbd4dfd71/processors/README.md#event-to-spanevent-bridge)):

- An SDK-based log processor that converts Event records to Span Events

>   and attaches them to the current span, whose behavior and configuration

Reviewer: This doesn't fit well with separately configured non-global tracer/logger providers. Logs will simply attach to the current span regardless of which tracer that span came from. In particular this means that tracers can no longer provide flexible utilities for automatically recording spans on their own. Either tracer providers must be given a logger provider or they must use the global one.

Author: > In particular this means that tracers can no longer provide flexible utilities for automatically recording spans on their own.

yeah, this can be a nice convenience function, Java has been considering adding something like this, but it's still under incubation: https://github.com/open-telemetry/opentelemetry-java/blob/main/api/incubator/src/main/java/io/opentelemetry/api/incubator/trace/ExtendedSpanBuilder.java#L37-L49

potentially this could be solved by having the Trace SDK depend on the Logs API (so at least we don't need to couple the APIs)

or better yet: https://github.com/open-telemetry/oteps/pull/165 😢

Reviewer 4: Why can't `LogRecord`s simply use the span/spancontext in their provided `Context`? IIRC it is from this very `Context` that their proto `trace_id` and `span_id` are populated from.

Reviewer: They can. The question is how to create that record in the first place when all you have is a span/tracer. Which logger (provider) do you use? Does it have to be the global one?

Reviewer 3: 🤔 if we passed Context with Span to logger.emit then log->spanEvent processor could attach log record as span even directly to the provided span

```java
class LogToSpanEventProcessor implements LogRecordProcessor {
   void onEmit(Context context, ReadWriteLogRecord logRecord) {
      // filter out records without event name
      ...

      // we need to identify if span is a real span or wrapped span context
      Span span = Span.fromContext(context);
      if (span.isRecording()) {
         span.addEvent(logToSpanEvent(logRecord));
      } else if (sameSpanContext(context, logRecord.getSpanContext()) and Span.current().isRecording()) {
         span.addEvent(logToSpanEvent(logRecord))
      }  
      ...
   }
}
```

If you want to attach log record to a certain span, you give log record a context with a span (and not just span context) - you no longer depend on the `span.current` limitations.

The log record will be sent wherever this span goes (global or local provider) and there is no coupling between SDKs. Am I missing some use-case or problem?

Reviewer 3: > They can. The question is how to create that record in the first place when all you have is a span/tracer. Which logger (provider) do you use? Does it have to be the global one?

any - global or local, but logger provider needs to be configured with that processor. It does not need to know about  tracing since all it deals with is a span instance.

Reviewer: Again, the starting point is the tracing SDK. The Python SDK lets you write this:

```python
with tracer.start_span(...):
    # span-ending exceptions inside here are recorded automatically
```

This needs to move from `span.record_exception` to using the logging API/SDK. How does the tracer do that? If it can only use the global logger provider, that seems like a problem. Otherwise they need some coupling.

Reviewer 3: this otep suggests that someone who used to write

```python
tracer = get_local_or_global_trace_provider().get("test")

with tracer.start_span(...) as span:
    ...
    span.record_exception(ex)
```

would instead (after a major version bump in that instrumentation) write

```python
tracer = get_local_or_global_trace_provider().get("test")
logger = get_local_or_global_log_provider().get("test")

with tracer.start_span(...) as span:
    ...
    logger.emit(severity=appropriateSeverity, "foo_operation_failed", exception=ex, context = get_context(span))
```
(there is a lot of convenience that should be one around that `logger.emit`)

and then users who want to return the old behavior back, would configure the log-> span event processor in the SDK (with declarative config or programmatically)

Reviewer: `with tracer.start_span` records unhandled exceptions by itself. There wouldn't usually be `span.record_exception(ex)` unless you also wanted to record a handled exception, which IIUC is already discouraged.

Reviewer 3: Great point, I keep forgetting that python records exceptions by default. I believe it means that python OTel SDK will keep doing `span.record_exception(ex)` for many years to come (until there is OTel SDK 2.0). 
Tracing SDK should not start logging and we should probably document that tracing SDK should not record exceptions without user asking for it. 

@trask do you think it would be useful to mention that when it comes to OTel tracing SDKs, if they ever have v2, they should stop calling into `span.RecordException` by default (if they do it today) and should not start calling into `logger.emit`

Reviewer: @lmolkova combined with the resistance to adopt an alternative like https://github.com/open-telemetry/opentelemetry-specification/issues/4429 what I'm hearing you saying is that conveniences which automatically record span-ending exceptions *anywhere* (whether in attributes, span events, or logs) are undesirable. Am I misunderstanding?

Reviewer 3: not really, I'm saying that the SDK should not do it by default.

Author: I'm resolving this discussion, but if I'm mistaken, please open a new discussion summarizing the remaining open discussion points, thanks!

# Span Event API deprecation plan

## Motivation

[OTEP 265: Event Vision](0265-event-vision.md) states that we intend to
deprecate Span Events in favor of (log-based) Events.

After further discussions, we are only planning to deprecate the Span Event
API, while retaining the ability to emit Span Events via the Logs API.

This achieves the primary goal of deprecation, which is to provide a single
consistent guidance to instrumentation authors to use the Logs API when
emitting events, while still supporting use cases that rely on Span Events
being emitted in the same proto envelope as their containing span.

## The plan

### In the proto

Stabilize (log-based) Events.

### In the Specification

> 1. Stabilize emitting exceptions and events via the Logs API.

Author: does https://github.com/open-telemetry/opentelemetry-specification/pull/4430/commits/8337009f077d9670677f523b6c3dc12bd14b6c41 help?

   This will allow recording exceptions and events using the Logs API,
   instead of recording them using the Span Event API.

   This can be done in parallel with 1.

2. Mark [Span RecordException](../specification/trace/api.md#record-exception)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that exceptions are recorded using the Logs API.

3. Mark [Span AddEvent](../specification/trace/api.md#add-events)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that events are recorded using the Logs API.

   This can be done in parallel with 2.

### Per API and SDK

> 1. Implement and stabilize emitting exceptions and events via the Logs API.

Reviewer 5: I am a bit unsure of "stabilize emitting exceptions via the Logs API" part... what does it mean? Do you envision that OTel Logs API will be exposing some helper to report exceptions, similar to span.RecordException?

Author: does https://github.com/open-telemetry/opentelemetry-specification/pull/4430/commits/8337009f077d9670677f523b6c3dc12bd14b6c41 help?

- In the instrumentation's current major version
  - It SHOULD continue to use these
- In the instrumentation's next major version
  - It SHOULD stop using these,
    instead recording additional details about spans (that don't need a timestamp)
    as spans attributes,
    and recording exceptions and events using the Logs API.
  - Users will be able to retain the prior telemetry output by opting in to
    - [Sending (log-based) Events as Span Events](#sending-log-based-events-as-span-events)

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance.

## Sending (log-based) Events as Span Events

### Via the SDK

There MUST be a way to send (log-based) Events as Span Events
for use cases that rely on Span Events being emitted in the
same proto envelope as their containing span, and for users
who need to preserve the prior behavior when updating to
new instrumentation.

This mechanism SHOULD be implemented as follows (see
[prototype](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/80adbe1cf8de647afa32c68f921aef2bbd4dfd71/processors/README.md#event-to-spanevent-bridge)):

> - An SDK-based log processor that converts Event records to Span Events

Reviewer 5: could you mention the this does require LogSDK (i.e A valid LoggerProvider) to be setup.

Reviewer 5: depending on the language and whether the language has declarative config, this may/may-not be code changes.

Author: > could you mention the this does require LogSDK (i.e A valid LoggerProvider) to be setup.

I'm not sure what this adds to the existing text? maybe you could make a specific suggestion?

> depending on the language and whether the language has declarative config, this may/may-not be code changes.

declarative config is covered in the next bullet

instrumentations that are using
[Span RecordException](../specification/trace/api.md#record-exception)
and [Span AddEvent](../specification/trace/api.md#add-events):

- In the instrumentation's current major version
  - It SHOULD continue to use these
- In the instrumentation's next major version
  - It SHOULD stop using these,
    instead recording exceptions and events using the Logs API.
  - Users will be able to retain the prior telemetry output by opting in to
    [Sending (log-based) Events as Span Events](#sending-log-based-events-as-span-events)
  - In case the instrumentation was using span events previously to record
    additional details about a span (i.e. details that don't need a timestamp),
    it SHOULD instead record these details as attributes on the span. See
    [semantic-conventions#2010](https://github.com/open-telemetry/semantic-conventions/issues/2010)
    and
    [opentelemetry-specification#4446](https://github.com/open-telemetry/opentelemetry-specification/issues/4446)
    for more details about this case.

Non-stable instrumentations SHOULD use their best judgement on whether to follow
the above guidance.

> ## Sending (log-based) Events as Span Events

Reviewer 6: @trask What was the response for [this comment](https://github.com/open-telemetry/opentelemetry-specification/pull/4430/files#r2054286087)?  I share a similar concern. Backend systems that only accept traces like zipkin, would now be losing visibility into exceptions without a similar functionality like this for recording exceptions.

Author: isn't that addressed by this section?

> There MUST be a way to send (log-based) Events as Span Events
for use cases that rely on Span Events being emitted in the
same proto envelope as their containing span, and for users
who need to preserve the prior behavior when updating to
new instrumentation.

Reviewer 6: If the expectation for this section is that when an exception is logged it automatically can be attached to an active span, then fine. I didn't think that was clear though.

Author: oh, I understand now, thanks! yes, that was the intention but was not what the doc said. I've addressed in https://github.com/open-telemetry/opentelemetry-specification/pull/4430/commits/bf68fa502da1277300d93d3b0982d5e3dedaa5cc

### Via the SDK

There MUST be a way to send (log-based) Events as Span Events
for use cases that rely on Span Events being emitted in the
same proto envelope as their containing span, and for users
who need to preserve the prior behavior when updating to
new instrumentation.

> This mechanism SHOULD be implemented as follows (see

Reviewer: Doing this via the SDK only makes it impossible for instrumentations to intentionally emit span events, even through the logging API. This essentially means that span events can't be recommended in semantic conventions, which goes against the motivation "supporting use cases that rely on Span Events being emitted in the same proto envelope as their containing span".

Here is an example of where this is a problem: https://github.com/open-telemetry/semantic-conventions/issues/2010#issuecomment-2747859183

Many people want to capture GenAI content in the same envelope as the span, which is pushing us towards using giant span attributes containing complex JSON, whereas span events would work much better.

It's impossible for the instrumentation to default to emitting span events if this can only be done in the SDK. But it would be quite easy for the API to contain a very minimal implementation of LoggerProvider which attaches a span event to the current span, without involving processors at all. This way instrumentations and users still use the logging API for everything, but there's a logging API that explicitly produces span events where desired. This makes it possible for instrumentations to use logs without requiring users to actually configure logs at all.

Reviewer 4: @alexmojaki Out of curiosity, if complex attributes were to be supported, would this solve this part?

> Many people want to capture GenAI content in the same envelope as the span, which is pushing us towards using giant span attributes containing complex JSON

Reviewer: 1. How will large complex attributes be handled? Is there a spec for configuring limiting the size of complex log bodies already? Is there a way to limit the size of individual nested values so that large prompts inside can be truncated while keeping the overall structure?
2. Some people *do* want to use log-events (like the spec is now) because this *does* have benefits, like handling long conversations better and not forcing everything into a single payload which might be too big. Ideally the instrumentation would allow choosing between putting the data in the span or in logs. But this makes the spec/semconv/implementation more complex. You could maybe define a generic format for converting between log-events and an attribute containing an array. But you can't define a logger or log processor that adds an item to a span attribute because that's not a supported operation. On the flip side, this PR already implies that there will be a generic component and specified format to convert log events to span events, in which case instrumentations and their specs don't really have to do anything. All semantic conventions involving log-events that are children of spans will automatically imply an equivalent for span events. But with a small tweak we could give instrumentations the option to do that explicitly.

Reviewer: In Python, the convention for instrumentations is to accept optional tracer/meter/logger providers but use the global ones by default. I'm assuming it's the same in other languages.

For instrumentations like the current GenAI spec where log-events only exist as children of spans in the same instrumentation, what should happen if no explicit providers are given, a global tracing provider has been configured, and no global logging provider has been configured?

1. Throw an error. I don't think this is an option since I think in general instrumentations aren't supposed to interfere in application behaviour.
2. Discard the data that would go into logs. This is a confusing user experience. At best you can warn the user and hope they see it. Even then it sounds annoying. They might not be using a backend that supports logs. The only recommendation which is expected to always be available is a logger which attaches events to the current span.
3. Default to a logger which attaches events to the current span, but only if logging isn't globally configured? This means that if they later globally configure logging for unrelated reasons, some of their data will unexpectedly move to a different signal.

I don't even know if any of the above (except discarding the data *without* warning the user) are even possible - is there an API for checking if global configuration has happened? Is it acceptable for instrumentations to do that?

Arguably the best option is for such instrumentations to default to using a logger which attaches to the current span, even if logging is globally configured. It would be a bit inconsistent/unconventional when compared to other instrumentations which default to the global provider, but that seems better than the other options. But this isn't possible if attaching span events is only allowed through the SDK.

Reviewer 7: > there a spec for configuring limiting the size of complex log bodies already?

PTAL https://github.com/open-telemetry/opentelemetry-specification/issues/4006. It is triaged as "community-feedback". But if I remember correctly there is the idea to put the "event payload" as attribute. If so then see the next answer. CC @lmolkova @trask @breedx-splk as my memory is be flaky and may be outdated :wink: 

> Is there a way to limit the size of individual nested values so that large prompts inside can be truncated while keeping the overall structure?

Yes. See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#logrecord-limits

Reviewer: > > How will large complex attributes be handled?
> 
> How do you solve it today for span events?

The data looks something like this:

```
[{role: user, content: "Please summarize this book: <wall of text>"}, {role: assistant, content: "Sure: <smaller wall of text>"}]
```

If each object in the array is a span event, then all the attributes are simple, and existing per-attribute limits work great. Even if you need complex attribute types inside the span events, a per-attribute limit will still work better at truncating just the biggest parts while leaving the other span events intact, compared to putting the whole array inside a single span attribute.

> > Is there a spec for configuring limiting the size of complex log bodies already?
> 
> PTAL #4006. It is triaged as "community-feedback". But if I remember correctly there is the idea to put the "event payload" as attribute. If so then see the next answer. CC @lmolkova @trask @breedx-splk as my memory is flaky and may be outdated 😉

https://github.com/open-telemetry/opentelemetry-specification/issues/4006#issuecomment-2072534773 mentions the key problem:

> [AnyValue](https://github.com/open-telemetry/opentelemetry-proto/blob/342e1d4c3a1fe43312823ffb53bd38327f263059/opentelemetry/proto/common/v1/common.proto#L28-L67) is a recursive data structure, and the common attribute limits were designed for a flat list of key value pairs.

There's a proposal there for how to deal with this, an objection to that proposal from you, and that's where the discussion of that point has stopped, roughly a year ago.

> > Is there a way to limit the size of individual nested values so that large prompts inside can be truncated while keeping the overall structure?
> 
> See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/sdk.md#logrecord-limits

That only applies to standard attribute types.

Reviewer 7: > The data looks something like this:

This looks like a "hack". I there any OTel semantic convention which tells you to do that? 

> if complex attributes were to be supported, would this solve this part?

If complex attributes were to be supported in spans and metrics then you would have with the same problem, right?

> That only applies to standard attribute types.

Where does this information comes from (I ask about the specification, not some specific implementation of SDK which may be not spec compliant)? Logs signal operates on log attributes which are a superset of standard attributes. See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-attributes.

Author: discussed with @alexmojaki and he is ok to resolve this discussion

OpenTelemetry zero-code distribution (if one exists for the language).

## Backcompat story for span-terminating exceptions

### Via the SDK

This mechanism SHOULD be implemented as follows:

- An SDK-based span processor that converts span-terminating exceptions
  recorded as span attributes into Span Events
  (with the option to either drop or retain the span-terminating exception
  span attributes).
- A standard way to add this span processor via declarative configuration
  (assuming its package has been installed).

This span processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language).

### Via the Collector

This mechanism SHOULD be implemented as follows:

- A Collector-based span processor that converts span-terminating exceptions
  recorded as span attributes into Span Events.

This log processor SHOULD be included in the standard
OpenTelemetry Collector Contrib distribution.

## Communication plan

Publish a blog post if/when this OTEP is accepted, giving readers a way to
provide feedback (e.g. pointing to a specification issue where we are
gathering feedback). The blog post should include the rationale for the

> decision to deprecate Span Events.

Reviewer 8: ```suggestion
decision to deprecate Span Event API.
```

Stabilize (log-based) Events.

### In the Specification

1. Introduce and stabilize the Span SetException API.

   This will allow recording span-terminating exceptions directly as span
   attributes, instead of recording them as Span Events.

   Stabilization will go through the normal process, requiring prototypes
   in at least three languages.

2. Stabilize the (log-based) Event API.

   This will allow recording events using the (log-based) Event API,
   instead of recording them using the Span Event API.

   This can be done in parallel with 1.

3. Mark [Span RecordException](../specification/trace/api.md#record-exception)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that span-terminating exceptions are recorded directly
   as span attributes via the new Span SetException API and
   that other exceptions are recorded using the (log-based) Event API.

4. Mark [Span AddEvent](../specification/trace/api.md#add-events)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that events are recorded using the (log-based) Event
   API.

   This can be done in parallel with 3.

> ### Per SDK

Reviewer 7: nit

```suggestion
### Per API and SDK
```

1. Implement and stabilize the Span SetException API and the (log-based)
   Event API.

2. Implement the two SDK-based backcompat stories below:

   - [Sending span-terminating exceptions as Span Events](#via-the-sdk)

>    - [Sending (log-based) Events as Span Events](#sending-log-based-events-as-span-events)

Reviewer 7: What about the same in the collector?

This mechanism SHOULD be implemented as follows (see
[prototype](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/80adbe1cf8de647afa32c68f921aef2bbd4dfd71/processors/README.md#event-to-spanevent-bridge)):

- An SDK-based log processor that converts Event records to Span Events
  and attaches them to the current span, whose behavior and configuration
  are defined in the OpenTelemetry Specification.
- A standard way to add this log processor via declarative configuration
  (assuming its package has been installed).

Users can add a batch log record processor and log exporter depending on
if they also want to export Event records as logs or not.

This log processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language).

## Communication plan

Publish a blog post if/when this OTEP is accepted, giving readers a way to
provide feedback (e.g. pointing to a specification issue where we are
gathering feedback). The blog post should include the rationale for the
decision to deprecate the Span Event API.

## Nice to haves

- Make [sending span-terminating exceptions as Span Events via the
  Collector](#via-the-collector)

>   a prerequisite for stabilizing Span SetException in SDKs.

Reviewer 7: Not sure if this is needed given it is already present in `### Via the Collector`.

# Span Event API deprecation plan

## Motivation

[OTEP 265: Event Vision](0265-event-vision.md) states that we intend to
deprecate Span Events in favor of (log-based) Events.

After further discussions, we are only planning to deprecate the Span Event
API, while retaining the ability to emit Span Events via the (log-based)
Event API.

This achieves the primary goal of deprecation, which is to provide a single
consistent guidance to instrumentation authors to use the Event API,
while still supporting use cases that rely on Span Events being emitted in the

> same proto envelope as their containing span.

Reviewer 9: I'm not against this idea of consolidating the two separate ways of doing the same thing (events within spans, events linked/correlated with span context), but given that the [Event API and SDK have been removed](https://github.com/open-telemetry/opentelemetry-specification/pull/4353), I don't think that the wording here is accurate. We've gone through significant effort to land on a logs api that also supports events, and we should probably be clearer about that here.

As currently phrased, I worry that a reader may come away looking for the event api, only to find that one doesn't actually exist.

Author: yeah, this is a good point, we probably need to live with the consequences of [OTEP 265](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/0265-event-vision.md) and just refer to this as the Logs API

## Span-terminating exceptions

As we look at the current usage of Span Events across OpenTelemetry
repositories, by far the most common use is for recording span-terminating

> exceptions.

Reviewer 10: It is technically possible today to record multiple exception span events on a span. Albeit this is probably an uncommon scenario.

That said, I think this proposal should address this scenario in some way. My suggestion is to clarify that multiple calls to the new `SetException` API will result in span exception attributes from the last call made.

Author: SetException would only be for a "span-terminating" exception. Exceptions that aren't span-terminating would now be emitted using the Logs API.

Reviewer 10: My comment was more a question of the behavior of multiple calls to `SetException`. Perhaps too in the weeds for the OTEP?

Author: oh, yeah, I've seen OTEPs go in both directions (high- and low- level details), I think I'd prefer to keep this one high-level, and only dive into details where we think there is contention

This OTEP proposes to record span-terminating exceptions directly as span
attributes (since there is only ever at most one of them on a span) instead

> of recording them via the (log-based) Event API, for the following reasons:

Reviewer 11: This creates a situation where exceptions are recorded as logs, except for exceptions that are so important that they fail the operation. In which case, these exceptions are omitted from the logging stream. That seems bad.

I agree that it's best to couple the SetException method to the span, but by default I feel that the exception should also be emitted as a log. 

If users really don't like the double reporting, perhaps they could disable this feature. But it would be very surprising to see some exceptions in your log stream, only to discover later that the most important exceptions are absent.

Reviewer: > by default I feel that the exception should also be emitted as a log.

By default logs aren't configured, so this is not possible. And if the default behaviour is emitting a log *if* logs are configured, then:

1. How? Do I pass a logger provider to the tracer provider? Or does it just get the global provider? If I have three tracer providers, do they all use the same logger provider by default?
2. If I start configuring logs for unrelated reasons, then suddenly new logs containing exceptions will appear. This sounds very surprising.

> But it would be very surprising to see some exceptions in your log stream, only to discover later that the most important exceptions are absent.

Backends can choose to unpack span-terminating exceptions into new logs if they feel this is the best UX. We tried doing this ourselves and quickly regretted it because of point 5 in https://github.com/open-telemetry/opentelemetry-specification/issues/4429. But there's no easy way to deal with a separate log when that's undesirable.

Reviewer 3: I believe @tedsuo proposes that instrumentation change current `span.recordException(ex)` to 

```java
span.setException(ex);
logger.atLevel(levelDependingOnTheContext, "error when doing things", additionalAttributes, ex);
```

I wish we had instrumentation API that would hide these (along with metrics) two calls.

I agree we should leave a room in this OTEP to record exceptions as attributes on spans and also as logs - they are not mutually exclusive.

Reviewer 11: If we are giving users the option of receiving SpenEvents as Logs, then we're already moving into a world where TracingProviders need to take a LogProvider as an argument, correct?

Reviewer 11: But in general, I don't want to create complicated SDK pipelines and circular dependencies between Providers. It feels like "the simplest thing that could work" would be to have two instrumentation calls: one to the logger to record the exception, and one to the span to end it. The exception would then continue to get associated with the span as a SpenEvent if you use that feature, or it would be recorded as a log with the associated SpanContext. Some information can also be duplicated between the log and the span attributes if that is helpful. That would be the most consistent and straightforward solution.

Author: resolving based on latest updates to the OTEP: https://github.com/open-telemetry/opentelemetry-specification/pull/4430#issuecomment-2734504077

- Span-terminating exceptions are (by their nature) recorded via tracing
  instrumentation and so there is a natural interest to keep them tightly
  coupled to the tracing (span) telemetry.

> - It is _much_ simpler to build a backcompat collector processor that moves

Reviewer 5: ```suggestion
- It is _much_ simpler to build a back-compat processor (in Sdk or in collector) that moves
```

Author: this is pretty specific to the collector, it's not necessarily hard to do in the SDK where you have the ability to access the current span in a log processor

Reviewer 5: aah. got it!

  span-terminating exceptions from span attributes to a Span Event,
  compared to buffering spans and moving these span-terminating exceptions
  over to the appropriate span.

## The plan

### In the proto

Stabilize (log-based) Events.

### In the Specification

> 1. Introduce and stabilize the Span SetException API.

Reviewer 11: Where is the SetException API defined? I don't see it in this doc.

Author: resolving since SetException is gone now in the latest revision

   This will allow recording span-terminating exceptions directly as span
   attributes, instead of recording them as Span Events.

   Stabilization will go through the normal process, requiring prototypes
   in at least three languages.

2. Stabilize the (log-based) Event API.

   This will allow recording events using the (log-based) Event API,
   instead of recording them using the Span Event API.

>    This can be done in parallel with 1.

Reviewer 9: I'm having the same reaction here as above, regarding the use of "Event API". I respect that we've tried to indicate that it's _part_ of the logs api by using `(log-based)` there, and I'm not trying to be pedantic nor difficult, but I don't think that we can/should refer to a single method (`setEventName()`) on an existing logs API as its own API.

Author: the term "Event API" is now removed everywhere

3. Mark
   [Span RecordException](../specification/trace/api.md#record-exception)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that span-terminating exceptions are recorded directly
   as span attributes via the new Span SetException API and
   that other exceptions are recorded using the (log-based) Event API.

4. Mark [Span AddEvent](../specification/trace/api.md#add-events)
   as [Deprecated](../specification/document-status.md#lifecycle-status),
   recommending instead that events are recorded using the (log-based) Event
   API.

   This can be done in parallel with 3.

### Per Instrumentation

For [stable](../specification/versioning-and-stability.md#stable)
instrumentations that are using
[Span RecordException](../specification/trace/api.md#record-exception)
and [Span AddEvent](../specification/trace/api.md#add-events):

- In the instrumentation's current major version
  - It SHOULD continue to use these
- In the instrumentation's next major version
  - It SHOULD stop using these,
    instead calling the new Span SetException API for span-terminating exceptions

>     and calling the (log-based) Event API for all other use cases.

Reviewer 3: based on the recent discussions in the GenAI SDK and spec, I'm suggesting to add another option (for instrumentations) that uses span attributes to log things that are used-to-be-span-events, but are not point-in-time telemetry. 

It's the same case as span-terminating-exceptions.
E.g. http request body content (OTel instrumentations don't emit them, but if they did, attribute might be a reasonable option instead of event).

So suggesting something like
```suggestion
    , calling AddAttribute on span (when recording additional details about spans that don't have timestamp) and calling the (log-based) Event API for all other use cases.
```

I don't think we need to talk about complex attributes for it since users can express span-event attributes as span attributes (primitives or arrays of primitives, or json string) similarly to the way we'll record terminating exception.

Reviewer 12: Could you be more clear about what this would look like? The first thing I jumped to is 'typed attribute' which in my mind is something like passing a 'complex' attribute to `SetAttribute`, but the 'complexity' would be defined as some sort of logical event... doesn't this get us back to the same place we were before, where instrumentors have to pick one or both?

Reviewer 3: Let's forget about complex/typed attributes for the sake of this discussion. 

Imagine I recorded something like a span event before. What do I do after deprecation:
- if it's a span-terminating exception -> record it with `AddException` on span - i.e. it will be recorded as `exception.*` attributes directly on span.
- if it's a point-in-time data potentially useful without a span -> use Events (aka Logs) API. E.g. rpc message events are in this category.

What if it's neither? The timestamp is not quite applicable and the data belongs on the span:
- Obvious example: error details (error without exception)
- Debatable examples: http request body content, database query response

There are probably some other things end-user applications attach as span events that are only important to them when attached to a span and their timestamp matches either start of end of the span.

So what I'm saying that these things can be added as span attributes (when it makes sense).
E.g.  if I record database query response as span event, I'd have something like

```yaml
span:
  span_name: "SELECT users"
  attributes:
    - db.query.text: "SELECT id, name, age FROM users"
    - ...
  events:
    event_name: db.query.response
    attributes:
      - db.query.response: "1, Alice, 25\n2, Bob, 42"
```

I can express it on span attributes as well with 

```yaml
span:
  span_name: "SELECT users"
  attributes:
    - db.query.text: "SELECT id, name, age FROM users"
    - ...
    - db.query.response: "1, Alice, 25\n2, Bob, 42"
```

-----

If I happen to emit event per row now

```yaml
span:
  span_name: "SELECT users"
  attributes:
    - db.query.text: "SELECT id, name, age FROM users"
    - ...
  events:
    - event_name: db.query.response.row
      attributes:
      - db.query.response.row: "1, Alice, 25"
    - event_name: db.query.response.row
      attributes:
      - db.query.response.row: "2, Bob, 42"
```

span attributes would be

```yaml
span:
  span_name: "SELECT users"
  attributes:
    - db.query.text: "SELECT id, name, age FROM users"
    - ...
    - db.query.response.rows: 
      - "1, Alice, 25", 
      - "2, Bob, 42"
```

TL;DR: this discussion is not related to complex attributes, it's about things that belong on the span and don't need a separate log record.

Reviewer 3: maybe something like this would be more clear

```suggestion
    , recording additional details about spans (that don't need timestamp) as spans attributes, and calling the (log-based) Event API for all other use cases.
```

Reviewer 12: Ah, ok, that's more clear yes. This seems like something that SemConv would need to handle, right?

Reviewer 3: yes, semconv would need to document if something should be captured as an event or an attribute. 
I'm also thinking about custom span events reported by user apps - they would need to make a choice when Span Event API is deprecated.  We'll try our best to provide the generic guidance in https://github.com/open-telemetry/semantic-conventions/issues/1855

for use cases that rely on Span Events being emitted in the
same proto envelope as their containing span, and for users
who need to preserve the prior behavior when updating to
new instrumentation.

This mechanism SHOULD be implemented as follows (see
[prototype](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/80adbe1cf8de647afa32c68f921aef2bbd4dfd71/processors/README.md#event-to-spanevent-bridge)):

- An SDK-based log processor that converts Event records to Span Events
  and attaches them to the current span, whose behavior and configuration
  are defined in the OpenTelemetry Specification.
- A standard way to add this log processor via declarative configuration
  (assuming its package has been installed).

Users can add a batch log record processor and log exporter depending on
if they also want to export Event records as logs or not.

This log processor SHOULD be included in the standard
OpenTelemetry zero-code distribution (if one exists for the language).

### Via the Collector

Note: this is a nice to have only and not required for any other part
of the Span Event API deprecation plan.

This mechanism MAY be implemented as follows:

> - A Collector-based processor that buffers spans and moves (log-based)

Reviewer 13: I know it is just a nice to have so not a blocker but how feasible is this processor? Technically it'd be feasible with a global map that any instance of the span processor could read from and any event/log pipeline can write to but would acceptable to the collector maintainers or is there another way I'm missing?

Author: I agree this would not be easy, it's similar to one of the motivating reasons behind the introduction of span-terminating exceptions above:

> - It is _much_ simpler to build a backcompat collector processor that moves span-terminating exceptions from span attributes to a Span Event, compared to buffering spans and moving these span-terminating exceptions over to the appropriate span.

is there any specific update you'd suggest that we make to the text in this section here?
